### PR TITLE
Fixes microdoses of drugs giving you traits forever

### LINKED ISSUE #36152: You can gain a permanent speedboost by smoking
Round ID: 84467

If you add a small enough ratio of speed boosting chemicals to a cigarette/cigar/e-cig, and smoke it for a second, the speed boost you get will never run out.

[Reproduction]:
Prepare chemicals that boost your movement speed (meth, ephedrine, whatever else).
Water it down a bit. Let's say 20% of the boosting mix, 80% of anything else. 
Suck the mix in something you can smoke.
Smoke it.
You're now supersonic forever. You can take off the smoke and still keep it.
Bonus points by adding morphine, meaning you can run ultrafast no matter what you wear or how much damage you take.
So you can run around in a hardsuit and DUFFLEBAG at light speed.

You can even do this! [Sonic the clown](https://i.imgur.com/mXhyCOt.gif)
And yes it sounds as annoying as you think.

I only found this working for movement speed. Health regen, stun decrease or whatever else don't stick after smoking it. 
The health analyzer shows the reagents never entering you. At a higher ratio of the mix, it _does_, and then it just works normally.

### PR DESCRIPTION
**Pull request by Supermichael777**
Ready
:cl:
fix: NT would like to remind all employees that micro-dosing of prohibited drugs is not recognized as effective by the surgeon general.
/:cl:

fixes: #36152 

Because while being an unstunable super-speed god is fun its also a serious bug (and I don't want to get banned 😁 ). Fun fact it would also work with mute toxin, you could perma-mute someone if you got the dose right.

### CODE REVIEW

**Comment by Supermichael777** [comment]
wait actualy this might cause a bug with brave bull hold on

**Comment by GuyonBroadway** [comment]
Turns out Homeopathy works on ss13

**Comment by Supermichael777** [comment]
At least i caught that before it became meta

**Comment by Supermichael777** [comment]
This also fixes the old on and off meth run you used to get with cigs.

File: code/modules/reagents/chemistry/reagents.dm [hunk 1]
```diff
@@ -61,6 +62,7 @@
 
 // Called when this reagent is first added to a mob
 /datum/reagent/proc/on_mob_add(mob/M)
+	mobeffectactive = TRUE
```
Comment (Cyberboss) [inline code/modules/reagents/chemistry/reagents.dm]: You need to make sure that everything that overrides on_mob_add also calls ..()

**Comment by yorii** [comment]
dlet dis

**Comment by Supermichael777** [comment]
I got warned by a trial-min and Don't exploit bugs is this weird pseudo rule.

**Comment by Cyberboss** [comment]
Did you check the overrides?

**Comment by Supermichael777** [comment]
yes this is ready

File: code/modules/reagents/chemistry/reagents.dm (hunk 1 above)
Comment (Supermichael777) [reply code/modules/reagents/chemistry/reagents.dm]: they all do

File: code/modules/reagents/chemistry/reagents.dm [hunk 2]
```diff
@@ -61,6 +62,7 @@
...
 	return
```
Comment (Cyberboss) [inline code/modules/reagents/chemistry/reagents.dm]: remove this return

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/reagents/chemistry/reagents.dm (hunk 1 above)
Comment (Supermichael777) [reply code/modules/reagents/chemistry/reagents.dm]: and they just need to not hard return all procs call their parent at the end if they havent 1. alredy manualy called it 2. returned a value 3. hit the end of proc/whatever() and auto-returned null

Comment (Cyberboss) [reply code/modules/reagents/chemistry/reagents.dm]: >all procs call their parent at the end if they havent

This is false

**Review by Cyberboss — CHANGES_REQUESTED** [review]
All the on_mob_add's in alcohol_reagents.dm don't call ..() as well as mimesbane

**Comment by Supermichael777** [comment]
that should be handled as long as they do not add a trait as the var will be removed along with the reagent datum. i will look over it a second time

**Comment by Cyberboss** [comment]
Still though, if you add behavior to a previously empty parent proc, you need to make sure all overrides now call it

**Comment by Supermichael777** [comment]
Okay that should be sorted. Those ones I changed to reaction_mob made more sense as they seemed to want to do that whenever the reagent was added rather than when the datum was created.

**Comment by optimumtact** [comment]
I don't understand this fix, can you talk it through to me? The bug to me seems that it's not calling on_mob_delete when it should

**Comment by Supermichael777** [comment]
Okay so its two things:

origaly it was an order of operations problem; mob add was being called last in the add chain so if you got a dose of less than .1 units you would get the effect forever because on mob delete was already called.

If you corected the order of operations the drugs gave you the effect for a single tick every life() process if you set up the microdose cig.  this was the behavior before they were swapped to this method back when mobs cleared themselves of these when they couldn't find a source.

My change fixes both the unusual behavior by making traits never activate if the reagent is qdeleting and never running the delete if the trait isn't active(this is probably unnecessary unless we get an on_mob_delete that runtimes).

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Add ..() to mimesbane

**Comment by Supermichael777** [comment]
99.9% sure that it inherits but okay

**Comment by ExcessiveUseOfCobblestone** [comment]
I would think it would be simpler to have a threshold within the add_reagents proc that doesn't call on_mob_add if the amount is below such than doing this.

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]

File: code/modules/reagents/chemistry/reagents/toxin_reagents.dm [hunk 3]
```diff
@@ -910,6 +910,8 @@
 
 /datum/reagent/toxin/mimesbane/on_mob_add(mob/living/L)
 	L.add_trait(TRAIT_EMOTEMUTE, id)
+	()..
```
[the remark this answers is not in the document]
Comment (optimumtact) [reply code/modules/reagents/chemistry/reagents/toxin_reagents.dm]: lmao

**Comment by ExcessiveUseOfCobblestone** [comment]
@Supermichael777 just fyi unless it says within the block that it inherits (or you don't ever define a proc specific to that subtype), it will not inherit.

Only you can ~~prevent codefires~~ tell it to inherit!

**Comment by AnturK** [comment]
This going anywhere ?

**Comment by optimumtact** [comment]
to the mooooooon

**Comment by Cyberboss** [comment]
m8

**Pull request closed without merging**